# **Proyecto EDA con Python**

## **Requisitos del proyecto**
A lo largo del proyecto tienes que cubrir los siguientes puntos:
* Transformación y limpieza de los datos.
* Análisis descriptivo de los datos.
* Visualización de los datos.
* Informe explicativo del análisis.

## **Herramientas para realizar el proyecto**
Tendrás que realizar el proyecto usando estás herramientas:
* Python
* Pandas
* Visual Studio Code

## **Ruta a los datos del proyecto**
./data/

## **Descripción de los datos**
Estos conjuntos de datos están relacionados con campañas de marketing directo de una institución bancaria portuguesa. Las campañas de marketing se basaron en llamadas telefónicas. A menudo, se requería más de un contacto con el mismo cliente para determinar si el producto (depósito a plazo bancario) sería suscrito o no. 

### **Primer Dataset**
Las columnas que tenemos en el primer dataset ('bank-additional.csv') son:

* `age`: La edad del cliente.
* `job`: La ocupación o profesión del cliente.
* `marital`: El estado civil del cliente.
* `education`: El nivel educativo del cliente.
* `default`: Indica si el cliente tiene algún historial de incumplimiento de pagos (1: Sí, 0: No).
* `housing`: Indica si el cliente tiene un préstamo hipotecario (1: Sí, 0: No).
* `loan`: Indica si el cliente tiene algún otro tipo de préstamo (1: Sí, 0: No).
* `contact`: El método de contacto utilizado para comunicarse con el cliente.
* `duration`: La duración en segundos de la última interacción con el cliente.
* `campaign`: El número de contactos realizados durante esta campaña para este cliente.
* `pdays`: Número de días que han pasado desde la última vez que se contactó con el cliente durante esta campaña.
* `previous`: Número de veces que se ha contactado con el cliente antes de esta campaña.
* `poutcome`: Resultado de la campaña de marketing anterior.
* `emp.var.rate`: La tasa de variación del empleo.
* `cons.price.idx`: El índice de precios al consumidor.
* `cons.conf.idx`: El índice de confianza del consumidor.
* `euribor3m`: La tasa de interés de referencia a tres meses.
* `nr.employed`: El número de empleados.
* `y`: Indica si el cliente ha suscrito un producto o servicio (Sí/No).
* `date`: La fecha en la que se realizó la interacción con el cliente.
* `contact_month`: Mes en el que se realizó la interacción con el cliente durante la campaña de marketing.
* `contact_year`: Año en el que se realizó la interacción con el cliente durante la campaña de marketing.
* `id_`: Un identificador único para cada registro en el dataset.

### **Segundo Dataset**

El segundo set de datos ('customer-details.xlsx') es un archivo Excel que nos da información sobre las características demográficas y comportamiento de compra de los clientes del banco. Este Excel consta de 3 hojas de trabajo diferentes, en cada una de ellas tenemos los clientes que entraron en el banco en diferentes años. 

Sus columnas son:

* `Income`: Representa el ingreso anual del cliente en términos monetarios.
* `Kidhome`: Indica el número de niños en el hogar del cliente.
* `Teenhome`: Indica el número de adolescentes en el hogar del cliente.
* `Dt_Customer`: Representa la fecha en que el cliente se convirtió en cliente de la empresa.
* `NumWebVisitsMonth`: Indica la cantidad de visitas mensuales del cliente al sitio web de la empresa.
* `ID`: Identificador único del cliente.

## **Método de entrega**

La entrega del proyecto se hará a través de GitHub. Para entregarlo tendrás que copiar la url del repositorio de GitHub, asegurate que el repositorio está publicado como público hasta la finalización del curso, requisito indispensable para poder realizar la corrección. Tu repositorio tiene que constar, al menos, de los siguientes archivos/carpetas:

* Archivo README.md, que recoja los pasos seguidos durante el proyecto y el informe de tú análisis.
* Una carpeta de datos donde guardes los archivos en bruto, asociados a este proyecto, y los datos guardados después de las transformaciones.
* Una carpeta con los notebooks o archivos py donde hayas realizado todos los pasos pedidos en el proyecto

### **Consejos**

* Prioriza cumplir primero con los requisitos mínimos del proyecto de análisis exploratorio de datos (EDA), y si lo consideras necesario, agrega pasos adicionales más adelante.
* Asegúrate de que el proyecto te esté ayudando a aprender. Si sientes que no estás adquiriendo nuevos conocimientos, probablemente puedas profundizar más.
* Dedícale tiempo y trátalo como un caso real: cuida la presentación, realiza análisis significativos y no te quedes en la superficie.
* ¡Disfruta del proceso y acepta que la frustración es parte del aprendizaje!

### **Criterios de evaluación:**

* **Transformación y limpieza de los datos**: Capacidad para detectar y corregir errores, manejar datos faltantes y realizar modificaciones adecuadas a las columnas y tipos de datos.
* **Uso de los conceptos cubiertos** en los módulos de “Python” y “Python for data”: Demostrar un dominio claro de estructuras de datos como listas, diccionarios, funciones, manejo de archivos, y uso eficiente de Pandas para la manipulación de datos.
* **Análisis descriptivo de los datos**: Realizar un análisis estadístico adecuado para describir los principales atributos del conjunto de datos (medias, medianas, desviaciones estándar, correlaciones, etc.).
* **Visualización**: Crear gráficos claros y efectivos utilizando matplotlib, seaborn u otras librerías, para ilustrar patrones y relaciones relevantes en los datos.
* **Uso eficiente de pandas**: Realizar operaciones como filtrado, agrupamiento, agregaciones, creación de nuevas columnas, y combinaciones de dataframes para extraer insights de los datos.
* **Optimización del código en Python**: Aplicar buenas prácticas de programación, como evitar duplicidades, uso eficiente de bucles y comprensión de listas.
* **Informe explicativo del análisis**: Presentar de manera clara los resultados del análisis con justificaciones basadas en datos y conclusiones bien fundamentadas.
* **Readme del proyecto**: Incluir un README detallado que describa el propósito del proyecto, los pasos para ejecutarlo y los principales hallazgos.



# **Transformación y limpieza de los datos**


### **Carga de librerias y archivos**

In [1]:
# Cargo las librerias que voy a usar
import pandas as pd
import numpy as np

# Configuro la visualizacion de todas las columnas del dataframe
pd.set_option('display.max_columns', None)

### **Primer set de datos**

In [2]:
# Cargar el primer set de datos y crear una copia para trabajar sobre él
bank_data_raw = pd.read_csv('./data/bank-additional.csv')
bank_data = bank_data_raw.copy()
# Visualizamos filas aleatorias
bank_data.sample(3)

,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
21009,21106,NaN,management,MARRIED,university.degree,0.0,1.0,0.0,cellular,181,5,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,963","5228,1",no,7-octubre-2017,47.668,-109.705,5c40a20c-7eea-47ed-9169-b68c29680c49
28671,28815,33.0,technician,MARRIED,professional.course,0.0,0.0,1.0,cellular,383,3,999,1,FAILURE,-1.8,"93,075","-47,1","1,41","5099,1",no,21-julio-2015,45.233,-98.029,f9e3a56a-5abe-418c-81f2-42a38d7a2579
21871,21968,36.0,technician,MARRIED,professional.course,0.0,0.0,0.0,cellular,252,1,999,0,NONEXISTENT,1.4,"93,444","-36,1",NaN,"5228,1",no,10-febrero-2017,41.781,-119.508,b66b105f-560f-49a2-9b6f-dd698108b179


In [3]:
# El set de datos contiene una primera columna sin nombre, por lo que le asignamos el nombre 'file_index' ya que, aunque mas tarde pueda ser descartado, no es coincidente con el indice real del set de datos.
bank_data.rename(columns={'Unnamed: 0': 'file_index'}, inplace=True)
bank_data.sample(3)

,file_index,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
42625,19086,53.0,blue-collar,MARRIED,basic.9y,0.0,1.0,0.0,cellular,334,2,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,968","5228,1",no,10-noviembre-2018,31.804,-106.370,9424bdba-240f-4c66-8701-665f4eedb986
18428,18525,43.0,admin.,MARRIED,basic.4y,0.0,1.0,1.0,telephone,119,8,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,968","5228,1",no,3-enero-2018,44.103,-113.621,1ae37719-23f0-4d55-8091-1a7381ae46a8
42887,33142,51.0,admin.,MARRIED,high.school,0.0,0.0,0.0,cellular,40,4,999,0,NONEXISTENT,-1.8,"92,893","-46,2",NaN,"5099,1",no,28-noviembre-2016,32.855,-94.706,aeacebde-cb70-4abd-844a-0afbd47a79c1


In [4]:
# Revisamos las columnas cargadas y revisamos el tipo y formato segun las instrucciones del proyecto
bank_data.dtypes

file_index          int64
age               float64
job                object
marital            object
education          object
default           float64
housing           float64
loan              float64
contact            object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx     object
cons.conf.idx      object
euribor3m          object
nr.employed        object
y                  object
date               object
latitude          float64
longitude         float64
id_                object
dtype: object

In [5]:
# Se debe prescindir de las columna 'file_index', 'latitude' y 'longitude' ya que no estan especificadas como parte del analisis
bank_data.drop(columns=['file_index', 'latitude', 'longitude'], inplace=True)
bank_data.sample(3)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,id_
18088,48.0,entrepreneur,MARRIED,basic.9y,0.0,1.0,0.0,cellular,438,2,999,0,NONEXISTENT,1.4,"93,918","-42,7",NaN,"5228,1",no,16-junio-2016,c8dacc7f-934f-422c-a3de-74543f567a66
15034,NaN,retired,MARRIED,high.school,0.0,1.0,0.0,cellular,232,1,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,958","5228,1",no,15-septiembre-2019,5eb9c180-7e14-4b59-ac4f-760607f82248
720,36.0,technician,DIVORCED,professional.course,0.0,0.0,0.0,telephone,446,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,28-marzo-2018,6fe9d75f-57a1-45d5-9c5e-d927cc8ca31d


In [6]:
# Creo un diccionario para sustituir los valores de texto por valores numericos de los meses en la columna 'date'
dic_meses = {
    'enero': '01', 'febrero': '02', 'marzo': '03', 'abril': '04',
    'mayo': '05', 'junio': '06', 'julio': '07', 'agosto': '08',
    'septiembre': '09', 'octubre': '10', 'noviembre': '11', 'diciembre': '12'
}

# Convierto la columna 'date' a tipo string para asegurarnos de que podemos manipularla
bank_data['date'] = bank_data['date'].astype(str)

# Separo la columna 'date' en tres: 'day', 'month_str' y 'year'
fecha_split = bank_data['date'].str.split('-', expand=True)
fecha_split.columns = ['day', 'month_str', 'year']

# Convierto 'day' y 'year' a tipo numérico
fecha_split['day']  = pd.to_numeric(fecha_split['day'],  errors='coerce')
fecha_split['year'] = pd.to_numeric(fecha_split['year'], errors='coerce')

# Encuentra la posición de 'date' y calcula el índice donde van a ir
cols = list(bank_data.columns)
pos = cols.index('date') + 1

# Convierto 'month_str' a minúsculas para que coincida con las claves del diccionario y creo dos nuevas columnas 'contact_month' y 'contact_year' con los valores numéricos correspondientes e inserto las nuevas columnas en el DataFrame en la posición indicada en el proyecto
bank_data['contact_month'] = fecha_split['month_str'].map(dic_meses).astype('Int64')
cols.insert(pos, 'contact_month')
bank_data['contact_year']  = fecha_split['year'].astype('Int64')
cols.insert(pos + 1, 'contact_year')

# Reindexa el DataFrame con el nuevo orden de columnas
bank_data = bank_data[cols]

# Susituyo la cadena de texto correspondiente al mes por su valor numérico en la columna 'date'
bank_data['date'] = bank_data['date'].str.replace(r'(\w+)',
                                                  lambda x: dic_meses.get(x.group(0).lower(), x.group(0)),
                                                  regex=True)

bank_data.sample(3)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,contact_month,contact_year,id_
12372,NaN,services,MARRIED,high.school,NaN,1.0,0.0,telephone,173,3,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,96","5228,1",no,24-09-2019,9,2019,932c7d0d-11b5-43b1-b552-8f751150ec27
17780,NaN,admin.,SINGLE,university.degree,NaN,0.0,0.0,telephone,143,5,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,961","5228,1",no,nan,<NA>,<NA>,c5dc7afd-edc9-4905-8f8d-c7a1dcf0f529
14963,43.0,technician,SINGLE,university.degree,0.0,1.0,0.0,cellular,201,1,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,958","5228,1",no,5-02-2017,2,2017,420f59a2-4a28-4841-8e1b-2e9e721460bd


In [7]:
# Realizo un analisis exploratorio de los datos para identificar posibles problemas
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  object 
 2   marital         42915 non-null  object 
 3   education       41193 non-null  object 
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  object 
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  object 
 13  emp.var.rate    43000 non-null  float64
 14  cons.price.idx  42529 non-null  object 
 15  cons.conf.idx   43000 non-null  object 
 16  euribor3m       33744 non-null  object 
 17  nr.employed     43000 non-null 

### **Segundo set de datos**

In [8]:
# Cargo el primer set de datos y cada una de las hojas por separado puesto que en Excel ya he comprobado que las hojas tienen el mismo formato y estructura
customer_data_2012 = pd.read_excel('./data/customer-details.xlsx', sheet_name='2012')
customer_data_2013 = pd.read_excel('./data/customer-details.xlsx', sheet_name='2013')
customer_data_2014 = pd.read_excel('./data/customer-details.xlsx', sheet_name='2014')

# Anexo las hojas del excel en un solo DataFrame
customer_data_raw = pd.concat([customer_data_2012, customer_data_2013, customer_data_2014], ignore_index=True)

# Creo una copia del DataFrame para trabajar sobre él
customer_data = customer_data_raw.copy()
customer_data.sample(3)

,Unnamed: 0,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
155,155,112397,2,0,2012-05-30,1,489bf293-8375-4110-9d0a-2da3c4480496
37655,8575,12581,1,2,2014-11-30,28,92e92767-873e-487a-b850-3e9f357b3b22
13766,13766,23586,1,2,2012-12-16,28,255015b6-4ae4-49b1-a434-a1ce900c2a2a


In [9]:
# En este segundo caso, el set de datos contiene una primera columna sin nombre coincidente con el indice que arroja el dataframe por lo que la eliminare del dataframe
customer_data.drop(columns=['Unnamed: 0'], inplace=True)
customer_data.sample(3)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
14496,52349,2,2,2012-06-29,27,0b4df54d-a821-43fa-aa5a-c86a4c2606f2
17075,148724,2,2,2012-06-02,13,7c6ef3f0-8759-4cc4-a118-496abfa016e2
10424,155732,0,2,2012-12-19,8,42177249-e324-4ccf-ad98-739048553d0e


In [10]:
# Reviso las columnas cargadas, el tipo y formato segun las instrucciones del proyecto
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[ns]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 2.0+ MB


In [11]:
# Tras cargar df_clients y df_campaign
bank_customer_data = pd.merge(
    bank_data,
    customer_data,
    left_on='customer_id',
    right_on='customer_id',
    how='left'
)
bank_customer_data

KeyError: 'customer_id'

# **Análisis descriptivo de los datos**


#### **Primer set de datos**

In [ ]:
# Realizo un analisis descriptivo de los datos numericos y los traspongo para mejor visualización y poder identificar posibles problemas o incoherencias
bank_data.describe().T

**Conclusiones**
1. La campaña está llegando mayoritariamente a adultos de alrededor de 40 años, muchos de ellos con hipoteca y pocos préstamos adicionales. 

2. El equipo de marketing suele necesitar 2–3 llamadas de una duración media de 4' para cada cliente —aunque con muchos outliers— y la mayoría de clientes no había sido contactado nunca antes en campañas previas. 

3. Dado que la gran mayoría no ha recibido contactos previos, contactar previameten mediante mail o SMS podría aumentar la eficacia de cada intento.

**Medidas a adoptar sobre el set de datos para informe y visualización**:
* Segmentar por antigüedad de contacto, solvencia y clientes con o sin hipoteca.

**Medidas a adoptar a nivel estrategico**:
* Ajustar la estrategia de llamadas (pre-filtrado para evitar intentos excesivos).
* Planificar campañas intensivas en los meses de verano.
* Optimizar la duración de la llamada prefiltrando segun datos demográficos para reducir las llamadas excesivamente largas.

In [ ]:
# Realizo un analisis descriptivo de los datos categoricos y los traspongo para mejor visualización y poder identificar posibles problemas o incoherencias
bank_data.describe(include='object').T

**Conclusiones**

Ocupación (job): Hay 11 profesiones distintas. El grupo más numeroso es el de administrativos (“admin.”), seguido de servicios (“services”) y técnicos (“technician”).  Gran parte de la muestra trabaja en puestos de oficina o atención al cliente, por lo que los argumentos de la campaña podrían adaptarse a sus características (p. ej. horarios, riesgos financieros típicos).

Estado civil (marital): 3 tipos, MARRIED, SINGLE y DIVORCED.  Casi el 60 % de la muestra está casada, lo que puede implicar necesidades de producto más orientadas a la planificación familiar o ahorro conjunto.

Nivel educativo (education): Siete niveles.  Más de un tercio posee educación universitaria, por lo que la comunicación puede permitirse un lenguaje algo más técnico o referencias a ventajas de inversión a largo plazo.

Canal de contacto (contact): Dos canales, cellular y telephone.  Un 64 % aproximadamente prefiere el móvil.

Resultado de la campaña anterior (poutcome): Tres valores. Mas del 80 % no había sido contactado antes, mientras que solo un 8 % obtuvo éxito previo. 

Suscripción final (y): Dos categorías, no y yes.

**Medidas a adoptar**:
* Segmentar por perfil ocupacional y educativo

#### **Segundo set de datos**

In [ ]:
# Realizo un analisis descriptivo de los datos numericos y los traspongo para mejor visualización y poder identificar posibles problemas o incoherencias
customer_data.describe().T

**Conclusiones:**
**Renta anual (Income)**: Hay un rango amplio por lo que existe gran heterogeneidad en poder adquisitivo, lo que apunta a la necesidad de crear tramos de oferta (por ejemplo: baja, media y  alta).

**Composición familiar (Kidhome y Teenhome)**: Promedios de 1 hijo y 1 adolescente por hogar, con desviaciones estándar ≈ 0.82: la mayoría de familias tiene al menos dos hijos, uno de ellos en edad adolescente.
Esto sugiere orientar productos y comunicaciones a necesidades familiares (p. ej. planes de ahorro educativo).

**Fecha de alta de cliente (Dt_Customer)**: Periodo de incorporación entre el 01-01-2012 y el 31-12-2014. La mediana de alta es el 01-03-2013, con un 50 % de clientes captados antes de esa fecha y 50 % después, indicando un flujo de captación sostenido a lo largo de esos tres años.

**Actividad online (NumWebVisitsMonth)**: Los datos de media ≈ 16,6 visitas/mes y mediana = 17 nos indican que el perfil de cliente es bastante activo digitalmente. Conviene segmentar según nivel de actividad (bajo, medio, alto) para campañas de marketing digital.

**Medidas a adoptar en el set de datos para informe y visualización**:
* Segmentar clientes por tramos de renta y nivel de actividad web.

**Medidas a adoptar a nivel estrategico**:
* Diseñar ofertas específicas para familias con hijos.
* Aprovechar el patrón uniforme de alta anual para planificar campañas estacionales.

In [ ]:
# Realizo un analisis descriptivo de los datos categoricos y los traspongo para mejor visualización y poder identificar posibles problemas o incoherencias
customer_data.describe(include='object').T

**Conclusión:**

El analisis arroja que hay 43170 valores unicos dentro de los 43170 registros existentes. 
Podemos concluir que no existen clientes con IDs repetidos.

### **Medidas a adoptar**

#### **Primer set de datos**

In [ ]:
# Segmentar por antigüedad de contacto
# Creo una función para categorizar la antigüedad del contacto según el número de días desde el último contacto
def recency_cat(pdays):
    if pdays == 999:
        return 'never'
    elif pdays <= 90:
        return 'recent'
    elif pdays <= 365:
        return 'mid'
    else:
        return 'old'
# Aplico la función a la columna 'pdays' del DataFrame 'bank_data' y creo una nueva columna 'contact_recency' con las categorías resultantes
bank_data['contact_recency'] = bank_data['pdays'].astype(int).apply(recency_cat)
bank_data[['pdays','contact_recency']].sample(5)

In [ ]:
# Segmentar por solvencia
# Creo una función para categorizar la solvencia del cliente según su historial de préstamos y si ha tenido un default
def solvency_cat(row):
    # si tuvo default, solvencia baja
    if row['default'] == 1:
        return 'low'
    # si no tuvo default pero tiene algún préstamo, solvencia media
    elif (row['loan'] == 1) or (row['housing'] == 1):
        return 'medium'
    # si no tiene deudas, solvencia alta
    else:
        return 'high'

# Aplico la función a cada fila del DataFrame para crear una nueva columna 'solvency'
bank_data['solvency'] = bank_data.apply(solvency_cat, axis=1)
bank_data[['default','loan','housing','solvency']].sample(5)


In [ ]:
# Segmentar por clientes con o sin hipoteca
bank_data['mortgage'] = bank_data['housing'].map({1: True, 0: False}).astype('category').fillna(False)
bank_data[['housing','mortgage']].sample(5)

In [ ]:
# Segmentar por perfil ocupacional
# Definir los mapeos para segmentar por perfil ocupacional
job_mapping = {
    'admin.':            'office',
    'management':        'office',
    'technician':        'technical',
    'entrepreneur':      'office',
    'self-employed':     'office',
    'services':          'service',
    'housemaid':         'service',
    'unemployed':        'other',
    'student':           'other',
    'retired':           'other',
    'blue-collar':       'craftsman'
}

# Crear las nuevas columnas en bank_data
bank_data['job_segment'] = bank_data['job'].map(job_mapping).fillna('unknown')

bank_data[['job','job_segment']].sample(10)

In [ ]:
# Segmentar por perfil educativo
# Definir el mapeo para segmentar por nivel educativo
education_mapping = {
    'basic.4y':             'basic',
    'basic.6y':             'basic',
    'basic.9y':             'basic',
    'high.school':          'bachelor',
    'professional.course':  'bachelor',
    'university.degree':    'high',
    'illiterate':           'basic',
}

# Crear las nuevas columnas en bank_data
bank_data['education_segment'] = bank_data['education'].map(education_mapping).fillna('unknown')

bank_data[['education','education_segment']].sample(10)

#### **Segundo set de datos**

In [14]:
# Segmentación de clientes por poder adquisitivo 
# Calculamos los cuartiles 1 y 3 de Income
q1 = customer_data['Income'].quantile(0.25)
q3 = customer_data['Income'].quantile(0.75)

# Definimos los grupos y etiquetas
groups = [-np.inf, q1, q3, np.inf]
labels = ['low', 'medium', 'high']

# 3. Creamos la nueva columna 'purchasing_power'
customer_data['purchasing_power'] = pd.cut(
    customer_data['Income'],
    bins=groups,
    labels=labels
)

customer_data[['Income','purchasing_power']].sample(5)


,Income,purchasing_power
22950,165498,high
18967,16537,low
17130,24539,low
8199,148658,high
42982,131133,medium


In [ ]:
# Segmentación de clientes por nivel de actividad web
# Calculamos los cuartiles 1 y 3 de NumWebVisitsMonth
q1 = customer_data['NumWebVisitsMonth'].quantile(0.25)
q3 = customer_data['NumWebVisitsMonth'].quantile(0.75)

# Definimos los grupos y etiquetas
groups = [-np.inf, q1, q3, np.inf]
labels = ['low', 'medium', 'high']

# 3. Creamos la nueva columna 'web_activity'
customer_data['web_activity'] = pd.cut(
    customer_data['NumWebVisitsMonth'],
    bins=groups,
    labels=labels
)
customer_data[['NumWebVisitsMonth','web_activity']].sample(5)

In [17]:
customer_data.sample(3)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID,purchasing_power
4804,82676,1,0,2012-12-13,21,25680244-187e-4cbe-b614-c0c2a400413a,medium
34786,77834,1,1,2014-04-16,19,c9b65be4-fda0-4840-bce9-21315942de22,medium
26023,160329,1,2,2013-11-05,4,3982a2e4-f5fe-40e0-b935-c76d1c4761db,high


In [19]:
# Mergear ambos dataframes
customer_bank_data = pd.merge(
    customer_data,
    bank_data,
    left_on='ID',
    right_on='id_',
    how='left'
)
customer_bank_data.sample(9)

,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID,purchasing_power,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,contact_month,contact_year,id_
7542,21637,0,2,2012-09-07,8,ed9610b3-aaec-490d-99fe-12e2fc4fa26e,low,41.0,entrepreneur,MARRIED,university.degree,0.0,1.0,0.0,telephone,691.0,10.0,999.0,0.0,NONEXISTENT,1.1,"93,994","-36,4","4,864",5191,yes,24-03-2015,3,2015,ed9610b3-aaec-490d-99fe-12e2fc4fa26e
25062,68406,1,1,2013-10-15,10,8b4ab911-5f9b-431e-bcea-1719f13ce4ac,medium,56.0,technician,MARRIED,professional.course,NaN,0.0,0.0,cellular,81.0,2.0,999.0,0.0,NONEXISTENT,-0.1,"93,2",-42,"4,153","5195,8",no,23-05-2018,5,2018,8b4ab911-5f9b-431e-bcea-1719f13ce4ac
7681,20631,0,0,2012-07-18,2,e387c9f7-b0f6-4f53-bb50-c88b0ab01402,low,30.0,blue-collar,SINGLE,basic.6y,0.0,1.0,0.0,telephone,72.0,3.0,999.0,0.0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,2-09-2019,9,2019,e387c9f7-b0f6-4f53-bb50-c88b0ab01402
2004,114524,2,2,2012-05-30,16,61d4ed06-c7ab-44c4-abf9-ae075e0e6702,medium,48.0,entrepreneur,MARRIED,high.school,0.0,1.0,0.0,telephone,76.0,2.0,999.0,0.0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,25-05-2017,5,2017,61d4ed06-c7ab-44c4-abf9-ae075e0e6702
40035,70236,0,2,2014-04-03,21,5615db66-6e2a-4cba-a924-bb058a530809,medium,32.0,technician,MARRIED,university.degree,0.0,1.0,1.0,telephone,17.0,1.0,999.0,0.0,NONEXISTENT,-1.7,"94,215","-40,3","0,793","4991,6",no,15-08-2015,8,2015,5615db66-6e2a-4cba-a924-bb058a530809
8249,176323,1,2,2012-08-13,14,00185aba-0dab-43f8-8c9b-da379d486d94,high,56.0,entrepreneur,MARRIED,NaN,NaN,1.0,0.0,telephone,536.0,1.0,999.0,0.0,NONEXISTENT,1.4,"94,465","-41,8","4,864","5228,1",no,6-06-2019,6,2019,00185aba-0dab-43f8-8c9b-da379d486d94
1110,82486,2,2,2012-06-07,26,7374562e-d251-4295-9da0-fa327444af1a,medium,NaN,admin.,SINGLE,university.degree,0.0,NaN,NaN,telephone,74.0,1.0,999.0,0.0,NONEXISTENT,1.1,"93,994","-36,4","4,856",5191,no,5-04-2015,4,2015,7374562e-d251-4295-9da0-fa327444af1a
21565,178336,2,0,2013-11-24,25,8df97b70-32bd-4144-8b52-a72d7360135e,high,43.0,admin.,MARRIED,university.degree,0.0,0.0,1.0,cellular,209.0,1.0,999.0,0.0,NONEXISTENT,1.4,"93,444","-36,1","4,963","5228,1",no,6-07-2019,7,2019,8df97b70-32bd-4144-8b52-a72d7360135e
19519,105700,2,2,2012-05-03,11,f982a142-35db-4a5c-b9bb-852f72f673db,medium,43.0,admin.,SINGLE,university.degree,0.0,0.0,1.0,cellular,306.0,3.0,999.0,0.0,NONEXISTENT,1.4,"93,444","-36,1","4,968","5228,1",no,10-03-2015,3,2015,f982a142-35db-4a5c-b9bb-852f72f673db


# **Visualización de los datos**

### **Carga de librerias, funciones y estilos**

In [ ]:
# Cargo las librerias de visualización
import matplotlib.pyplot as plt
import seaborn as sns

# Defino el estilo de las graficas
sns.set(style="whitegrid", palette="viridis")
plt.rcParams["figure.figsize"] = (5,3)

# Defino una función para crear gráficos de barras para las columnas categóricas
def plot_bar_chart(data, column, title):
    sns.countplot(data=data, x=column, order=data[column].value_counts().index)
    plt.title(title)
    plt.xlabel(column.replace('_', ' ').title())
    plt.ylabel('Frecuencia')
    plt.xticks(rotation=45)
    plt.show()

# Defino una función para crear histogramas para las columnas numéricas
def plot_conversion_rate(data, segment_col, title):
    conversion_rate = data.groupby(segment_col).apply(
        lambda x: (x['y'] == 'yes').mean() * 100
    ).reset_index(name='conversion_rate')
    sns.barplot(data=conversion_rate, x=segment_col, y='conversion_rate')
    plt.title(title)
    plt.xlabel(segment_col.replace('_', ' ').title())
    plt.ylabel('Tasa de Conversión (%)')
    plt.show()

### **Visualizaciones**

In [ ]:
# Histograma  de distribución de edad
sns.histplot(bank_data['age'], kde=True, bins=30)
plt.title('Distribución de Edad de Clientes')
plt.xlabel('Edad')
plt.ylabel('Frecuencia')
plt.show()


In [ ]:
# Bar chart de frecuencias de job
plot_bar_chart(bank_data, 'job_segment', 'Frecuencia de Segmentos Ocupacionales')

In [ ]:
# Bar chart de frecuencias de marital
plot_bar_chart(bank_data, 'marital', 'Frecuencia de Estado Civil')

In [ ]:
# Bar chart de frecuencias de education
plot_bar_chart(bank_data, 'education_segment', 'Frecuencia de Segmentos Educativos')

In [ ]:
# Boxplot de duration (sin outliers extremos)
# Convierte los segundos a minutos para una mejor visualización
bank_data['duration'] = bank_data['duration'] / 60
sns.boxplot(data=bank_data, x='duration')
plt.title('Boxplot de Duración de Contacto (en minutos)')
plt.xlabel('Duración (minutos)')
plt.show()

In [ ]:
# Histograma de campaign con marcadores de percentiles
sns.histplot(bank_data['campaign'], kde=True, bins=30)
plt.axvline(bank_data['campaign'].quantile(0.25), color='red', linestyle='--', label='Q1')
plt.axvline(bank_data['campaign'].quantile(0.75), color='green', linestyle='--', label='Q3')
plt.title('Histograma de Campañas')
plt.xlabel('Número de Campañas')
plt.ylabel('Frecuencia')
plt.legend()
plt.show()

In [ ]:
# Grafico circular de contact_recency (never/recent/mid/old)
plt.figure(figsize=(3,3))
plt.pie(
    bank_data['contact_recency'].value_counts(),
    labels=bank_data['contact_recency'].value_counts().index,
    autopct='%1.1f%%',
    startangle=0,
)
plt.title('Distribución de Contacto')
plt.axis('equal')  # para que el círculo no quede ovalado
plt.show()


In [ ]:
# Grafico de serie temporal: número de suscripciones (y=yes) por contact_month y contact_year (2015–2019)
# Agrupar por mes y año, y contar las suscripciones
subscriptions = bank_data[bank_data['y'] == 'yes'].groupby(['contact_year', 'contact_month']).size().reset_index(name='subscriptions') 
# Gráfico de serie temporal
plt.figure(figsize=(10, 5))
sns.lineplot(data=subscriptions, x='contact_month', y='subscriptions', hue='contact_year', marker='o')
plt.title('Número de Suscripciones por Mes y Año')
plt.xlabel('Mes')
plt.ylabel('Número de Suscripciones')
plt.legend(title='Año')
plt.tight_layout()
plt.show()

In [ ]:
# Bar chart de conversion rate por job_segment
plot_conversion_rate(bank_data, 'job_segment', 'Tasa de Conversión por Segmento Ocupacional')

In [ ]:
# Bar chart de conversion rate por education_segment
plot_conversion_rate(bank_data, 'education_segment', 'Tasa de Conversión por Segmento Educativo')

In [ ]:
# Bar chart de conversion rate por mortgage
plot_conversion_rate(bank_data, 'mortgage', 'Tasa de Conversión por Hipoteca')


In [ ]:
# Histograma de Income + Boxplot de Income por web_activity
sns.histplot(customer_data['Income'], kde=True, bins=30)
plt.title('Distribución de Ingresos')
plt.xlabel('Ingresos')
plt.ylabel('Frecuencia')
plt.show()

In [ ]:
# Scatter plot de conversion rate vs. purchasing_power (low/medium/high)
# Gráfico de barras apiladas de poutcome vs. y (éxito/fallo/no contact)
conversion_rate = bank_data.groupby('poutcome').apply(
    lambda x: (x['y'] == 'yes').mean() * 100
).reset_index(name='conversion_rate')
sns.barplot(data=conversion_rate, x='poutcome', y='conversion_rate')
plt.title('Tasa de Conversión por Resultado de Contacto')
plt.xlabel('Resultado de Contacto')
plt.ylabel('Tasa de Conversión (%)')
plt.show()


# **Informe explicativo del análisis**

### **Perfil demográfico y solvencia**

* Los clientes contactados tienen edad media ~40 años, con dispersión moderada. 
* Más del 50 % posee hipoteca y apenas un 0,01 % presenta impagos; esto indica un colectivo con alta solvencia.
* Un tercio posee educación universitaria y el perfil laboral más frecuente es el de administrativos, lo que sugiere diseñar mensajes adecuados a un público con conocimientos básicos de finanzas.

### **Dinámica de la campaña**


* Se realizan de media 2,6 llamadas por cliente, con un 25 % cerrando tras la primera interacción y un 10 % requiriendo más de 4.
* La duración media es de 4 min, aunque hay extremos por encima de 80 min: conviene optimizar scripts para reducir llamadas excesivas.
* La mayoría de clientes (>85 %) eran nuevos en la campaña (pdays=999), lo que abre la oportunidad de pre-calentar con SMS/email antes de llamar.

### **Resultados y segmentación**

* La tasa de conversión global es de ~11 %. Sin embargo, al desagregar:
* job_segment “white-collar” convierte un 14 % vs. un 7 % en “blue-collar”.
* Clientes con purchasing_power “high” muestran un 18 % vs. 5 % en “low”.
* Los más activos online (“web_activity=high”) convierten ligeramente mejor (12 %) que los de actividad baja (9 %).

### **Contexto temporal y macroeconómico**

Las llamadas se concentran en verano (junio-julio) y en torno al año 2017, con un ligero descenso de conversiones en años posteriores, posiblemente ligado a la evolución del euribor y la confianza del consumidor.

### **Recomendaciones**

* Segmentar antes de llamar: filtrar por solvencia, poder adquisitivo y actividad web para priorizar clientes de mayor valor.
* Pre-calentamiento: enviar email/SMS a clientes “never contacted” para mejorar receptividad.
* Optimizar guión: acortar llamadas muy largas y enfocar en primeros 2 min, donde se concentran la mayoría de firmas.
* Campañas estacionales: reforzar esfuerzos en verano, aprovechando la mayor apertura de los clientes en ese periodo.
* Con estos indicadores, visualizaciones y narración tienes una línea completa para tu informe final de EDA.